src/services/TestResultsService.ts

```diff
@@ -343,34 +348,46 @@ export class TestResultsService {
     const adrTestTypeIds = ["41", "64", "65", "66", "67", "95", "102", "103", "104"];
     return adrTestTypeIds.includes(testType.testTypeId);
   }
-  public getMostRecentExpiryDateOnAllTestTypesByVin(vin: any) {
-    let maxDate = new Date(1970, 1, 1);
+  public getMostRecentExpiryDateOnAllTestTypesByVin(vin: string): Promise<IMaxDates> {
+    const maxDate = new Date(1970, 1, 1);
+    const maxDatesForTestTypes: IMaxDates = {maxDateForAllTestTypes: maxDate, maxDateForAnnualTestTypes: maxDate};
+
     return this.getTestResults({
       vin,
       testStatus: TEST_STATUS.SUBMITTED,
       fromDateTime: new Date(1970, 1, 1),
       toDateTime: new Date()
     })
         .then((testResults) => {
-          const filterTestTypes: any[] = [];
+          const filteredTestTypes: {allTestTypes: any[]; annualTestTypes: any[]} = {allTestTypes: [], annualTestTypes: []};
           testResults.forEach((testResult: { testTypes: any; vehicleType: any; vehicleSize: any; vehicleConfiguration: any; noOfAxles: any; }) => {
-            testResult.testTypes.forEach((testType: { testExpiryDate: string; }) => {
-              if (testType.testExpiryDate) {
-                filterTestTypes.push(testType);
+            testResult.testTypes.forEach((testType: { testExpiryDate: string; testCode: string; }) => {
+              // prepare a list of annualTestTypes and allTestTypes with expiry.
+              if (TestResultsService.isValidTestCodeForExpiryCalculation(testType.testCode.toUpperCase()) && testType.testExpiryDate) {
+                filteredTestTypes.annualTestTypes.push(testType);
+              } else if (testType.testExpiryDate) {
+                filteredTestTypes.allTestTypes.push(testType);
               }
             });
           });
-          return filterTestTypes;
+          return filteredTestTypes;
         }).then((testTypes) => {
-          testTypes.forEach((testType) => {
-            if (dateFns.isAfter(testType.testExpiryDate, maxDate) && testType.testExpiryDate) {
-              maxDate = testType.testExpiryDate;
+          // fetch maxDate for allTestTypes
+          testTypes.allTestTypes.forEach((testType) => {
```

Comment:
Can simplify this using dateFns.max(...testTypes.allTestTypes.map(testType => testType.testExpiryDate)).